src/Interpreters/JoinExpressionActions.cpp

```diff
@@ -0,0 +1,514 @@
+#include <Interpreters/JoinExpressionActions.h>
+#include <algorithm>
+#include <cstdint>
+#include <ranges>
+#include <stack>
+#include <string_view>
+#include <unordered_set>
+#include <Core/Block.h>
+#include <boost/noncopyable.hpp>
+#include <Functions/isNotDistinctFrom.h>
+#include <Common/HashTable/Hash.h>
+
+
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionsLogical.h>
+#include <Functions/FunctionsComparison.h>
+#include <Common/logger_useful.h>
+#include <fmt/ranges.h>
+
+#include <Interpreters/ActionsDAG.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int UNKNOWN_IDENTIFIER;
+}
+
+std::string_view toString(JoinConditionOperator op)
+{
+    switch (op)
+    {
+        case JoinConditionOperator::Equals: return "=";
+        case JoinConditionOperator::NullSafeEquals: return "<=>";
+        case JoinConditionOperator::Less: return "<";
+        case JoinConditionOperator::LessOrEquals: return "<=";
+        case JoinConditionOperator::Greater: return ">";
+        case JoinConditionOperator::GreaterOrEquals: return ">=";
+        case JoinConditionOperator::And: return "AND";
+        case JoinConditionOperator::Or: return "OR";
+        case JoinConditionOperator::Unknown: break;
+    }
+    throw Exception(ErrorCodes::LOGICAL_ERROR, "Illegal value for JoinConditionOperator: {}", static_cast<Int32>(op));
+}
+
+
+String toString(const BitSet & bitset)
+{
+    std::string str;
+    boost::to_string(bitset.bitset, str);
+    return str;
+}
+
+int BitSet::findFirstSet() const
+{
+    if (bitset.none())
+        return -1;
+
+    size_t pos = bitset.find_first();
+    size_t maxPos = pos;
+
+    while (pos != Base::npos)
+    {
+        maxPos = pos;
+        pos = bitset.find_next(pos);
+    }
+
+    if (maxPos > std::numeric_limits<int>::max())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "BitSet is too large");
+
+    return static_cast<int>(maxPos);
+}
+
+size_t BitSet::hashImpl() const
```

Comment:
I don't really like using BitSet as a key of unordered_map.
Now, it is used in two places: 
* dynamic programming (maybe we can use size_t instead of Set; it can work only for a few tables anyway
* to calculate names; I am sure it can be done without it

So, maybe we can remove it.